// @microsoft/generator-sharepoint/lib/generators/onPremComponent/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk = require("chalk");
const path = require("path");
const BaseGenerator_1 = require("../../common/BaseGenerator");
const Utils = require("../../common/utilities");
const OnPremWebpart = require("../onPremWebpart");
// tslint:disable-next-line:no-any
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
// tslint:disable-next-line:no-any
function defineOptions(generator) {
    generator.option('componentType', {
        description: `The type of component. Currently "webpart" is the only option.`,
        type: String
    });
    OnPremWebpart.defineOptions(generator);
}
exports.defineOptions = defineOptions;
/**
 * This class selects between multiple different types of components for instantiation.
 */
class ComponentSelectorGenerator extends BaseGenerator_1.BaseGenerator {
    constructor() {
        super(...arguments);
        this.friendlyName = 'Component Selection';
    }
    initializing() { }
    prompting() {
        return this.prompt([
            {
                type: 'list',
                name: 'componentType',
                default: 'webpart',
                when: () => !this.options.componentType,
                message: 'Which type of client-side component to create?',
                choices: [
                    {
                        name: 'WebPart',
                        value: 'webpart'
                    }
                ]
            }
        ]).then((answers) => {
            const componentType = answers.componentType || this.options.componentType;
            const commonOptions = {
                'skip-install': this.options['skip-install'],
                framework: this.options.framework,
                componentName: this.options.componentName,
                componentDescription: this.options.componentDescription,
                creatingSolution: this.options.creatingSolution
            };
            if (componentType === 'webpart') {
                const wpOptions = commonOptions;
                OnPremWebpart.composeWith(this, wpOptions);
            }
            else {
                this.log(chalk.red(`Invalid componentType: "${componentType}"`));
            }
        });
    }
    configuring() { }
    writing() { }
    install() { }
    end() { }
}
exports.ComponentSelectorGenerator = ComponentSelectorGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = ComponentSelectorGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
