// @microsoft/generator-sharepoint/lib/generators/extension/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk = require("chalk");
const path = require("path");
const BaseGenerator_1 = require("../../common/BaseGenerator");
const Utils = require("../../common/utilities");
const FieldCustomizer = require("../fieldCustomizer");
const CommandSet = require("../commandSet");
const ApplicationCustomizer = require("../applicationCustomizer");
// tslint:disable-next-line:no-any
function defineOptions(generator) {
    FieldCustomizer.defineOptions(generator);
    CommandSet.defineOptions(generator);
    generator.option('extensionType', {
        description: `The type of extension. Currently "ApplicationCustomizer", "FieldCustomizer", or "ListViewCommandSet"`,
        type: String
    });
}
exports.defineOptions = defineOptions;
// tslint:disable-next-line:no-any
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
/**
 * This class selects between multiple different types of extensions for instantiation.
 */
class ExtensionSelectorGenerator extends BaseGenerator_1.BaseGenerator {
    // tslint:disable-next-line:no-any
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Extension Selection';
    }
    initializing() { }
    prompting() {
        return this.prompt([
            {
                type: 'list',
                name: 'extensionType',
                default: 'ApplicationCustomizer',
                when: () => !this.options.extensionType,
                message: 'Which type of client-side extension to create?',
                choices: [
                    {
                        name: 'Application Customizer',
                        value: 'ApplicationCustomizer'
                    },
                    {
                        name: 'Field Customizer',
                        value: 'FieldCustomizer'
                    },
                    {
                        name: 'ListView Command Set',
                        value: 'ListViewCommandSet'
                    }
                ]
            }
        ]).then((answers) => {
            const extensionType = answers.extensionType || this.options.extensionType;
            const commonOptions = {
                'skip-install': this.options['skip-install'],
                framework: this.options.framework,
                componentName: this.options.componentName,
                componentDescription: this.options.componentDescription,
                creatingSolution: this.options.creatingSolution
            };
            if (extensionType === 'FieldCustomizer') {
                const fieldCustomizerOptions = commonOptions;
                FieldCustomizer.composeWith(this, fieldCustomizerOptions);
            }
            else if (extensionType === 'ListViewCommandSet') {
                const commandSetOptions = commonOptions;
                CommandSet.composeWith(this, commandSetOptions);
            }
            else if (extensionType === 'ApplicationCustomizer') {
                const appCustomizerOptions = commonOptions;
                ApplicationCustomizer.composeWith(this, appCustomizerOptions);
            }
            else {
                this.log(chalk.red(`Invalid extensionType: "${extensionType}"`));
            }
        });
    }
    configuring() { }
    writing() { }
    install() { }
    end() { }
}
exports.ExtensionSelectorGenerator = ExtensionSelectorGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = ExtensionSelectorGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
