// @microsoft/generator-sharepoint/lib/generators/applicationCustomizer/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const Utils = require("../../common/utilities");
const BaseExtension = require("../extension/BaseExtensionGenerator");
// tslint:disable-next-line:no-any
function composeWith(base, options) {
    Utils.compose(path.basename(__dirname), base, options);
}
exports.composeWith = composeWith;
// tslint:disable-next-line:no-any
function defineOptions(generator) {
    BaseExtension.defineOptions(generator, 'Application Customizer');
}
exports.defineOptions = defineOptions;
class ApplicationCustomizerGenerator extends BaseExtension.BaseExtensionGenerator {
    // tslint:disable-next-line:no-any
    constructor(args, options) {
        super(args, options);
        this.friendlyName = 'Application Customizer';
        this.codeName = 'ApplicationCustomizer';
        this.allowedFrameworks = [];
        defineOptions(this);
    }
    initializing() { }
    prompting() {
        return super.prompting();
    }
    configuring() {
        super.configuring();
        this.ensureDependencyGroup('applicationCustomizer');
        this.serveJson.addConfiguration(this.context.componentNameCamelCase, {
            pageUrl: 'https://contoso.sharepoint.com/sites/mySite/SitePages/myPage.aspx',
            customActions: {
                [this.context.componentId]: {
                    location: 'ClientSideExtension.ApplicationCustomizer',
                    properties: {
                        testMessage: 'Test message'
                    }
                }
            }
        });
    }
    writing() {
        super.writing();
    }
    install() {
        super.install();
    }
    end() { }
}
exports.ApplicationCustomizerGenerator = ApplicationCustomizerGenerator;
// Yeoman expects exports of this type, which overrides the TS exports above
// Ensure that anything marked as export above is also listed here
module.exports = ApplicationCustomizerGenerator;
module.exports.defineOptions = defineOptions;
module.exports.composeWith = composeWith;
