// @microsoft/generator-sharepoint/lib/common/PackageSolutionJsonManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const YeomanConfiguration_1 = require("./YeomanConfiguration");
const JsonManager_1 = require("./JsonManager");
class PackageSolutionJsonManager extends JsonManager_1.JsonManager {
    static load(filepath, fs) {
        if (this._instance) {
            throw new Error(`Cannot create a second instance of PackageSolutionJsonManager`);
        }
        let data;
        try {
            data = fs.readJSON(filepath);
        }
        catch (error) {
            data = undefined;
        }
        if (!data) {
            data = {
                '$schema': 'https://dev.office.com/json-schemas/spfx-build/package-solution.schema.json'
            };
        }
        PackageSolutionJsonManager._instance = new PackageSolutionJsonManager(filepath, fs, data);
    }
    static reset() {
        PackageSolutionJsonManager._instance = undefined;
    }
    static get instance() {
        return this._instance;
    }
    getSkipFeatureDeployment() {
        try {
            return this.data.solution.skipFeatureDeployment;
        }
        catch (error) {
            return false;
        }
    }
    setUpSolution(solutionContext) {
        const data = this.data;
        data.solution = {
            name: `${solutionContext.libraryName}-client-side-solution`,
            id: solutionContext.libraryId,
            version: '1.0.0.0'
        };
        if (YeomanConfiguration_1.YeomanConfiguration.environment !== 'onprem') {
            data.solution.includeClientSideAssets = true;
        }
        if (solutionContext.skipFeatureDeployment) {
            data.solution.skipFeatureDeployment = solutionContext.skipFeatureDeployment;
        }
        data.paths = {
            zippedPackage: `solution/${solutionContext.libraryName}.sppkg`
        };
    }
    addExtensionFeature(featureId) {
        const solution = this.data.solution;
        if (!solution) {
            return;
        }
        if (!solution.features) {
            solution.features = [];
        }
        const feature = {
            title: 'Application Extension - Deployment of custom action.',
            description: 'Deploys a custom action with ClientSideComponentId association',
            id: featureId,
            version: '1.0.0.0',
            assets: {
                elementManifests: [
                    'elements.xml'
                ]
            }
        };
        solution.features.push(feature);
    }
}
exports.PackageSolutionJsonManager = PackageSolutionJsonManager;
