// @microsoft/generator-sharepoint/lib/common/YeomanConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable-next-line:typedef
const _strings = {
    libraryName: 'libraryName',
    framework: 'framework',
    version: 'version',
    libraryId: 'libraryId',
    environment: 'environment'
};
/**
 * This class wraps an untyped property bag with types for
 * the generator.
 */
class YeomanConfiguration {
    static setStore(yeomanConfiguration) {
        YeomanConfiguration._store = yeomanConfiguration;
    }
    static get libraryName() {
        return YeomanConfiguration._store.get(_strings.libraryName);
    }
    static set libraryName(libraryName) {
        YeomanConfiguration._store.set(_strings.libraryName, libraryName);
    }
    static get version() {
        return YeomanConfiguration._store.get(_strings.version);
    }
    static set version(version) {
        YeomanConfiguration._store.set(_strings.version, version);
    }
    static get libraryId() {
        return YeomanConfiguration._store.get(_strings.libraryId);
    }
    static set libraryId(libraryId) {
        YeomanConfiguration._store.set(_strings.libraryId, libraryId);
    }
    static get environment() {
        return YeomanConfiguration._store.get(_strings.environment);
    }
    static set environment(environment) {
        YeomanConfiguration._store.set(_strings.environment, environment);
    }
}
exports.YeomanConfiguration = YeomanConfiguration;
